### PR TITLE
More test cases and fix indexing bugs and update lessons

### LINKED ISSUE #215: [PE-D] markpresent all do not show results immediately.
Entered the command `viewlsn grp/G01 lsn/2-1` `markpresent all` as shown below.

Success message was shown in command result box but attendances in 2-1 did not change to true.

-------------
Labels: `severity.High` `type.FunctionalityBug`
original: samlsm/ped#6

### PR DESCRIPTION
**Pull request by xinyee20**
for markpresent, markabsent, flagatt, unflagatt, setscore, addscore, subscore command
close #215

### CODE REVIEW

File: src/test/java/team/serenity/logic/commands/studentinfo/UnflagAttCommandTest.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    @Test
+    public void constructor_nullParameter_throwsNullPointerException() {
+        assertThrows(NullPointerException.class, () -> new UnflagAttCommand((Index) null));
+        assertThrows(NullPointerException.class, () -> new UnflagAttCommand((Student) null));
+    }
+
+    @Test
+    public void execute_unflagStudent_success() throws CommandException {
+        ModelStubWithStudentsFlagged modelStub = new ModelStubWithStudentsFlagged();
+        Student toUnflagAtt = new Student("Aaron Tan", "A0123456U");
```
Comment (Nijnxw) [inline src/test/java/team/serenity/logic/commands/studentinfo/UnflagAttCommandTest.java]: Instead of manually creating the student, perhaps you can use`StudentBuilder` to build the default student to show that a default student passes the test successful?

Same idea for the rest of the test files :)

File: src/test/java/team/serenity/logic/commands/studentinfo/UnflagAttCommandTest.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+        CommandResult commandResult = new UnflagAttCommand(toUnflagAtt).execute(modelStub);
+        assertEquals(String.format(UnflagAttCommand.MESSAGE_SUCCESS, toUnflagAtt),
+                commandResult.getFeedbackToUser());
+    }
+
+    @Test
+    public void execute_wrongName_throwsCommandException() throws CommandException {
+        ModelStubWithStudentsFlagged modelStub = new ModelStubWithStudentsFlagged();
+        Student wrongNameOne = new Student("Aaron", "A0123456U");
```
Comment (Nijnxw) [inline src/test/java/team/serenity/logic/commands/studentinfo/UnflagAttCommandTest.java]: Similar to my previous comment on manually creating students, perhaps you could use `StudentBuilder` to build a default student, and use `.withName()` method to create the "wrong" student name.

Rationale: the default student gives a positive test case, so the reason for negative test case is due to the "wrong" student name. Clearer and better to understand! 

Same idea for the rest of the test files :)

File: src/test/java/team/serenity/logic/commands/studentinfo/SubScoreCommandTest.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    public void constructor_nullParameter_throwsNullPointerException() {
+        assertThrows(NullPointerException.class, () -> new SubScoreCommand((Index) null, 1));
+        assertThrows(NullPointerException.class, () -> new SubScoreCommand((Student) null, 1));
+    }
+
+    @Test
+    public void execute_subScoreOutOfRange_failure() throws CommandException {
+        ModelStubWithStudentsWithScores modelStub = new ModelStubWithStudentsWithScores();
+        Student toSubScore = new Student("Aaron Tan", "A0123456U");
+        int subScoreOutOfRange = 4;
```
Comment (Nijnxw) [inline src/test/java/team/serenity/logic/commands/studentinfo/SubScoreCommandTest.java:36]: If I'm not wrong, the range that score can take in is `0-5`?
Maybe you could include other invalid scores such as negative integers, non-integers and integers larger than 5 to cover all the partitions?

File: src/test/java/team/serenity/logic/commands/studentinfo/SubScoreCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        CommandResult commandResult = new SubScoreCommand(toSubScore, validSubScore).execute(modelStub);
+        assertEquals(String.format(SubScoreCommand.MESSAGE_SUCCESS, toSubScore, expectedScore),
+                commandResult.getFeedbackToUser());
+    }
+
+    @Test
+    public void execute_studentAbsent_throwCommandException() throws CommandException {
+        ModelStubWithStudentsAbsent modelStub = new ModelStubWithStudentsAbsent();
+        Student toSubScore = new Student("Aaron Tan", "A0123456U");
+        int validSubScore = 1;
```
Comment (Nijnxw) [inline src/test/java/team/serenity/logic/commands/studentinfo/SubScoreCommandTest.java:59]: **Suggestion:**
Realised you defined `validSubScore` multiple times.

Perhaps you want to create a `VALID_SUB_SCORE` constant in the `TypicalStudentInfo` class for use in this and other files? Easier to reference and change if we change the range of valid scores :)

Similar tip for your other test files variables!!

File: src/test/java/team/serenity/logic/commands/studentinfo/AddScoreCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    @Test
+    public void execute_wrongIndex_throwsCommandException() {
+        ModelStubWithIndexWithScore modelStub = new ModelStubWithIndexWithScore();
+        Index wrongIndex = Index.fromOneBased(Integer.parseInt("2"));
+        int validScore = 1;
+        AddScoreCommand addScoreCommand = new AddScoreCommand(wrongIndex, validScore);
+
+        assertThrows(CommandException.class, String.format(MESSAGE_INVALID_PERSON_DISPLAYED_INDEX,
```
Comment (Nijnxw) [inline src/test/java/team/serenity/logic/commands/studentinfo/AddScoreCommandTest.java]: I think `MESSAGE_INVALID_PERSON_DISPLAYED_INDEX` is AB3 reference? Maybe change the name of the constant to match our case of invalid student index?

**Review by Nijnxw — COMMENTED** [review]
Overall the code quality for the student info commands looks better!! 
Test cases are well-written too, just some small tips to make your test cases better :)

File: src/test/java/team/serenity/logic/commands/studentinfo/AddScoreCommandTest.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+class AddScoreCommandTest {
+    @Test
+    public void constructor_nullParameter_throwsNullPointerException() {
+        assertThrows(NullPointerException.class, () -> new AddScoreCommand((Index) null, 1));
+        assertThrows(NullPointerException.class, () -> new AddScoreCommand((Student) null, 1));
+    }
+
+    @Test
+    public void execute_addScoreOutOfRange_failure() throws CommandException {
```
Comment (neoruien) [inline src/test/java/team/serenity/logic/commands/studentinfo/AddScoreCommandTest.java]: Instead of execute_addScoreOutOfRange_failure(), maybe it should be named as execute_addScoreOutOfRange_throwsCommandException()

File: src/test/java/team/serenity/logic/commands/studentinfo/AddScoreCommandTest.java [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        int originalScore = 3;
+        int expectedScore = originalScore + validAddScore;
+
+        CommandResult commandResult = new AddScoreCommand(toAddScore, validAddScore).execute(modelStub);
+        assertEquals(String.format(AddScoreCommand.MESSAGE_SUCCESS, toAddScore, expectedScore),
+                commandResult.getFeedbackToUser());
+    }
+
+    @Test
+    public void execute_studentAbsent_throwCommandException() throws CommandException {
```
Comment (neoruien) [inline src/test/java/team/serenity/logic/commands/studentinfo/AddScoreCommandTest.java]: Tiny naming convention error: throwsCommandException() instead of throwCommandException()

File: src/test/java/team/serenity/logic/commands/studentinfo/UnflagAttCommandTest.java (hunk 1 above)
Comment (xinyee20) [reply src/test/java/team/serenity/logic/commands/studentinfo/UnflagAttCommandTest.java]: The current implementation of StudentBuilder fails the test case as the name of the Student does not correspond to name of any student in the lesson created by lesson builder, so I did not use it haha Do you think it will be better to change the default student name for Student Builder (not sure whether will fail test case elsewhere) or leave my current implementation as it is?

Comment (xinyee20) [reply src/test/java/team/serenity/logic/commands/studentinfo/UnflagAttCommandTest.java]: oh I shall change the way the student is built and infuse in StudentBuilder like 'new StudentBuilder().withName("Aaron Tan").withId("A0123456U").build();' I think that should do the trick:)

File: src/test/java/team/serenity/logic/commands/studentinfo/UnflagAttCommandTest.java (hunk 2 above)
Comment (xinyee20) [reply src/test/java/team/serenity/logic/commands/studentinfo/UnflagAttCommandTest.java]: yup okay good idea!

File: src/test/java/team/serenity/logic/commands/studentinfo/UnflagAttCommandTest.java (hunk 1 above)
Comment (xinyee20) [reply src/test/java/team/serenity/logic/commands/studentinfo/UnflagAttCommandTest.java]: > The current implementation of StudentBuilder fails the test case as the name of the Student does not correspond to name of any student in the lesson created by lesson builder, so I did not use it haha Do you think it will be better to change the default student name for Student Builder (not sure whether will fail test case elsewhere) or leave my current implementation as it is?

I have decided to change the default student name to the correct one since it seems to make more sense that the default student build is a valid student. I fix some areas to pass the test cases too, they were pretty minor I realise, so that's good. I will proceed to use the default student build by `StudentBuilder`, thanks for the suggestion! It was a good point:)

File: src/test/java/team/serenity/logic/commands/studentinfo/AddScoreCommandTest.java (hunk 6 above)
Comment (xinyee20) [reply src/test/java/team/serenity/logic/commands/studentinfo/AddScoreCommandTest.java]: Ah yes, thanks for noticing, I will make the changes:)

File: src/test/java/team/serenity/logic/commands/studentinfo/AddScoreCommandTest.java (hunk 5 above)
Comment (xinyee20) [reply src/test/java/team/serenity/logic/commands/studentinfo/AddScoreCommandTest.java]: Yes, good point, thanks for pointing it out:)

File: src/test/java/team/serenity/logic/commands/studentinfo/SubScoreCommandTest.java (hunk 3 above)
Comment (xinyee20) [reply src/test/java/team/serenity/logic/commands/studentinfo/SubScoreCommandTest.java:36]: yes, range that score can take in is 0-5, in this case the subScoreOutOfRange is an int score that will be subtracted from the original score to result in the score going out of the range. In the case of this test case, the original score is 3 so 4 covers the case where the subScore is greater than the score. The invalid scores for non-integers test case is covered under SubScoreCommandParserTest, I will add in test case for negative integers.

File: src/test/java/team/serenity/logic/commands/studentinfo/AddScoreCommandTest.java (hunk 7 above)
Comment (xinyee20) [reply src/test/java/team/serenity/logic/commands/studentinfo/AddScoreCommandTest.java]: yes, i will change, sorry for the grammar mistake

File: src/test/java/team/serenity/logic/commands/studentinfo/SubScoreCommandTest.java (hunk 4 above)
Comment (xinyee20) [reply src/test/java/team/serenity/logic/commands/studentinfo/SubScoreCommandTest.java:59]: sure, I will do this in the next iteration of edits!

**Review by ryanlimjr — APPROVED** [review]
LGTM
